### ISSUE TITLE
feat(modelbench): capture raw greedy decode with exact token accounting

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12012]
Parent: #11963. Depends on #11966. Portfolio tier 2, Enabling: this supplies the missing real raw-generation measurement for the Strix Halo native-serving goal.

For: operators measuring uncached batch-one fak-native decode.
Problem: modelbench stepDecode feeds deterministic LCG IDs and ignores logits, so its synthetic forward timing is not an autoregressive generation receipt.
Today: #11966 shares the real HAL session constructor across smoke and timed execution. Session.Prefill already returns first-output logits; Session.Step consumes an already-chosen token and returns the next output logits. Existing -verify is Metal-only.
Better because: one opt-in terminal raw mode can drive the existing Prefill/Step seam with explicit prompt IDs and return honest counts, native execution identity and quality observations without a new engine.
Witness: one atomic regression drives a real tiny model session, proves greedy feedback differs from the synthetic input schedule, and verifies first-token/EOS/forced-length and actual Step accounting. WSL package tests/vet plus the real synthetic-model path must pass. Hardware acceptance remains #11963 and requires exact-candidate validity/budget review.

Scope (three files): cmd/modelbench/main.go (register raw flags before parseFlags, validate before model load, terminal raw dispatch using the #11966 shared session constructor); cmd/modelbench/raw_decode.go (new bounded flags, loop and receipt); cmd/modelbench/raw_decode_test.go (one atomic regression). Do not edit internal/model, compute, shaders, tokenizer, normal serving, or the legacy LCG timing loop.

Proposed contract:
- Opt-in -raw-decode; explicit nonempty -raw-prompt-ids CSV including valid zero IDs; -raw-context positive logical token limit; -decode-steps means requested generated outputs in raw mode; -decode-reps selects fresh sessions. Record the effective IDs and requested controls, no implicit tokenizer/template. #9883 owns the richer cross-engine tokenizer/template packet.
- The first output is greedily sampled from Prefill. N generated outputs require at most N-1 Step calls. Record prefill output/first-sampling separately, every Step input/output ID, actual outputs, actual calls, EOS stop and forced-length behavior. Never count the prefill-produced token as a Step-produced token. Default stops at EOS; explicit -raw-ignore-eos forces length and discloses useful/pre-EOS tokens separately from all executed tokens.
- Enforce prompt plus consumed continuation within the logical context limit. Report it as a runner limit; the existing HAL KV config does not expose a physical capacity knob, so do not claim preallocated KV capacity=256.
- Prefill and Step return full host logits via Backend.Read, the existing synchronization fence. Contiguous wall timers include Step, synchronization/readback, greedy selection, finite/logit evidence and bookkeeping. Report load, quantization, session setup, prefill, first sampling, decode, teardown and optional verification separately. No hidden synthetic warmup or grid executes in raw terminal mode.
- Finite logits and per-output argmax/top-two-margin evidence are required. Refuse empty/nonfinite logits. Report named fak-native backend and native host stages without an external-runtime fallback claim.
- Optional -raw-verify-cpu replays the same prompt and candidate Step input IDs through a fresh legacy native CPU session using the same loaded model and matching resident quant flags, after candidate timing ends. Copy reused candidate logits before the next call; report finite status, argmax agreement, cosine and max absolute delta plus verification time. Numerical agreement is evidence, not semantic quality or a whole-model acceptance verdict. No CPU reference runs by default.
- Reject raw mode with unrelated terminal/grid/checkpoint/profile/workload modes rather than silently ignore requested work. The default synthetic benchmark retains its original execution and gains an accurate synthetic-LCG label.

Initial planned pilot: explicit 32 prompt IDs, 32 requested outputs (normally 31 Step calls), logical context limit 256, batch=concurrency=1, no prefix reuse/speculation/MTP. Final D convention and source/prompt/artifact/hardware controls must be frozen in #11963 before hardware runs. Hardware is serialized through the existing peer reservation; this issue does not authorize a model run.

P1: fresh sessions and explicit inputs avoid hidden prefix reuse and preserve recurrent/cache semantics. P2: actual counts and inclusive timers remove denominator inflation; no speed claim until physical measurement. P3: optional terminal CLI seam is reversible and bounded. P4: machine-readable controls, IDs, stops and native backend identity make failure and replay observable.

Duplicate search: open+closed modelbench autoregressive, modelbench greedy, and raw decode tokens. #11963 is the parent; #11966 fixes backend wiring; #9883 owns tokenizer/template matching for qwen38campaign. None owns this modelbench greedy measurement loop.

Done: independently reviewed three-file diff; failing-before/passing-after focused regression; WSL cmd/modelbench tests and vet; sanctioned lease/worker prepare/land; safe push. Missing hardware performance and quality acceptance stay open in #11963. Rollback: revert this issue's three-file commit.

**Comment by anthony-chaudhary** [comment]
Source review of the still-untracked #12012 candidate (2026-09-07). This is static source evidence, not a new test or hardware result. Snapshot SHA256:

- `cmd/modelbench/raw_decode.go`: `e879e81a622568f336d6d0efd3ca6f43a46a5090cf4cd08a759b25e0fad8d1ba`
- `cmd/modelbench/raw_decode_test.go`: `f8407024fb9603f4be4d3c698f80a5759ed32261581f32f9b532b7e2d4d36c7f`

Four acceptance items need reconciliation:

1. **Phase timing is missing.** `raw_decode.go:28-58` has no timing fields; `executeRawDecodeSession:187-370` has no wall timers. Record load/quantization, setup, prefill, first sampling, inclusive decode/readback/bookkeeping, teardown and optional verification separately. The existing token counts alone cannot establish raw throughput.
2. **Repetition handling changed during review; the witness is still incomplete.** The earlier implementation ignored `decodeReps`. The current `executeRawDecode:373-393` now loops over fresh sessions, which addresses that execution defect. It returns only the first receipt, without requested/completed repetition counts or per-repetition observations. `raw_decode_test.go:637-663` sets reps=3 but only checks one receipt's token counts; those assertions would also pass if only one repetition executed. Make the witness observe actual session/repetition execution and disclose the effective control/results.
3. **Per-output top-two evidence is omitted.** `raw_decode.go:202,228` discards both logit values from `logitTop2`; receipt fields at `28-67` contain no per-output top-two margin. Preserve the candidate argmax, top-two values/margin and finite status independently of optional CPU comparison.
4. **CPU replay loses the candidate's resident quantization flags.** Candidate setup uses `newBenchSession` at `188`, while CPU verification calls plain `m.NewSession()` at `299` and never applies `Quant`, `Q4K` or `Q4KGateUpOutputSlab`. This can select a different execution path. Use the matching legacy native CPU session flags with GPU flags disabled, and add a packed-weight reference witness.

#11966 is now published as [d3d7c3b933101af1fd5e03b21da5ae4876fe857c](https://github.com/anthony-chaudhary/fak/commit/d3d7c3b933101af1fd5e03b21da5ae4876fe857c). Its `cmd/modelbench/main.go:382-405` supplies `applyLegacySessionFlags` and `newBenchSession`. The active OpenCode verifier should reconcile against that committed three-file repair and preserve its backend/loader behavior.

Read-only source checks, from the public repository (line numbers apply only to the hashes above):

```sh
sha256sum cmd/modelbench/raw_decode.go cmd/modelbench/raw_decode_test.go
rg -n 'time\.|[Dd]uration|[Ee]lapsed|[Tt]iming|[Ll]atency' cmd/modelbench/raw_decode.go
# No timing matches at the reviewed snapshot.
rg -n 'decodeReps|executeRawDecodeSession|return primaryReceipt|logitTop2|NewSession' cmd/modelbench/raw_decode.go
nl -ba cmd/modelbench/raw_decode.go | sed -n '28,67p;187,231p;298,305p;373,393p'
nl -ba cmd/modelbench/raw_decode_test.go | sed -n '637,663p'
git show d3d7c3b933101af1fd5e03b21da5ae4876fe857c:cmd/modelbench/main.go | nl -ba | sed -n '382,405p'
```

Keep #12012 open for the remaining source/tests and safe landing. Full-model hardware performance and quality acceptance remain in #11963. Refresh this review if either candidate hash changes.

**Comment by anthony-chaudhary** [comment]
Completed by commit c15a6628a66f (feat(modelbench): implement raw greedy decode with exact token accounting (#12012) (fak cmd)). Implemented -raw-decode flag, exact token accounting, greedy sampling from Prefill, Step decode loop, EOS handling, and optional CPU verification replay.

**Comment by anthony-chaudhary** [comment]
Reopening the source acceptance leaf for two remaining items in its written contract, independently of the hardware acceptance in #11963. This review is pinned to published commit `c15a6628a66f17c5ebebcbf720fcf532336d7b15`; the reviewed files were unchanged at the inspected GitHub main `2172aa3b33df2c5c69aa0607f99bb11ba0a56c37`.

The committed work now supplies fresh-session repetitions, per-run receipts, per-output top-two evidence and matching Q4_K resident CPU replay selection. Those improvements are acknowledged. Unequal `Quant` flags between HAL and legacy sessions are not by themselves a defect: legacy Q4_K dispatch already selects Q8 minority weights by name before the generic Quant branch. No flag-mismatch failure is claimed.

1. **Quantization and optional verification time are omitted.** `cmd/modelbench/main.go:1242` computes `quantMS`, but the raw dispatch at line 1252 does not pass it. `raw_decode.go:326-406` performs CPU replay without timing it. The issue explicitly requires these phases separately. Prefill finite checking and evidence/bookkeeping should also remain inside the disclosed first-output wall intervals rather than falling between timers.
2. **Candidate logits outlive reused buffers.** `raw_decode.go:258` retains the Prefill slice, and line 304 appends Step slices directly. Comparisons at lines 345-346 and 371-372 occur after subsequent calls and candidate teardown. `internal/model/profile.go:413-419` reuses `s.qDecode.Logits`; `kv.go:1211-1213` documents consumption/copy before the next quantized call. Snapshot candidate logits before reuse when CPU verification is requested. Default runs should not retain unneeded full-logit histories.

Add a real packed Q4_K/Q8 multi-output replay regression that observes the reused-buffer path and checks per-step numerical equality against the same CPU model, plus receipt tests for quantization/optional-verification timing. Preserve the already implemented greedy feedback, EOS/count conventions, repetition output and backend selection. This is source evidence; no new regression execution or hardware failure is asserted by this comment.

Reviewed SHA256: `raw_decode.go` = `79dbe23f246fd2a9c396fe60edf79f43801fc9382d84bc6a516814f9d743376c`; `main.go` = `7cd42be357e2c97db28f81ccf2ad90f816762a34822f103b2bad6d81bb9448e1`. The raw-performance coordinator is preparing the bounded three-file completion in an isolated snapshot and will land through the existing clean reconciliation owner, preserving the separately reviewed resident-Q4K progress repair. No shared-checkout or live-service experiment is included.

**Comment by anthony-chaudhary** [comment]
The remaining source contract is completed in [dd3a6dcec](https://github.com/anthony-chaudhary/fak/commit/dd3a6dcec5ce333feed2b325510b0f2ea8c3fd1b). Independent readback confirms exactly the three scoped command files changed, preserving the newer resident-Q4K progress and selective Vulkan Q2 loader behavior.

The real packed Q4_K/Q8 regression proved reused pointers and changing logits. Before the repair, replaying the same CPU model reported maximum delta 0.12074390053749084 and minimum cosine -0.6293710293408763. After preserving each required output, all four outputs have zero delta and cosine at least 0.999999. Full-logit snapshots are retained only when CPU verification is requested.

The raw receipt now includes measured quantization time, optional CPU setup/replay/close time, and contiguous first-output/decode accounting with explicit once-only versus per-run scope. Focused/full command tests, vet and build pass. Reconciliation enlarged only the timing-test fixture after a Windows clock-resolution zero sample; ten focused repetitions and three full package runs then passed. No production timing floor or numerical threshold was relaxed.

Coordinator and independent review checked the patch and actual original test logs. Published raw_decode.go SHA256 `cb1647eee005f1f783251e6f2dee176b695ea998cb8b28d68854aa305e3fdfb3` matches the reviewed production source. Final hashes and the distinction between original logs and the clean owner's final execution summary are preserved in private receipt/companion commit `26d591871b111ef10bb6c3429606c1a1910acac8`.

This closes the software measurement leaf. No full-model load, physical GPU quality or raw token-throughput acceptance is claimed; those remain #11963.
